akka-cluster-typed/src/test/scala/akka/cluster/typed/ActorSystemSpec.scala

```
@@ -4,28 +4,56 @@
 
 package akka.cluster.typed
 
+import java.nio.charset.StandardCharsets
+
 import scala.concurrent.Future
 import scala.concurrent.Promise
 import scala.concurrent.duration._
 import scala.util.control.NonFatal
 
 import akka.Done
 import akka.actor.CoordinatedShutdown
+import akka.actor.ExtendedActorSystem
 import akka.actor.InvalidMessageException
 import akka.actor.testkit.typed.scaladsl.TestInbox
 import akka.actor.testkit.typed.scaladsl.LogCapturing
 import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorRefResolver
 import akka.actor.typed.ActorSystem
 import akka.actor.typed.Behavior
 import akka.actor.typed.PostStop
 import akka.actor.typed.scaladsl.adapter._
 import akka.actor.typed.scaladsl.Behaviors
+import akka.serialization.SerializerWithStringManifest
 import com.typesafe.config.ConfigFactory
 import org.scalatest._
 import org.scalatest.concurrent.Eventually
 import org.scalatest.concurrent.ScalaFutures
 import org.scalatest.time.Span
 
+object ActorSystemSpec {
+
+  class TestSerializer(system: ExtendedActorSystem) extends SerializerWithStringManifest {
+    // Reproducer of issue #24620, by eagerly creating the ActorRefResolver in serializer
+    private val actorRefResolver = ActorRefResolver(system.toTyped)
+
+    def identifier: Int = 47
+    def manifest(o: AnyRef): String =
+      "a"
+
+    def toBinary(o: AnyRef): Array[Byte] = o match {
+      case TestMessage(ref) => actorRefResolver.toSerializationFormat(ref).getBytes(StandardCharsets.UTF_8)
+    }
+
+    def fromBinary(bytes: Array[Byte], manifest: String): AnyRef = manifest match {
+      case "a" => TestMessage(actorRefResolver.resolveActorRef(new String(bytes, StandardCharsets.UTF_8)))
+    }
```

I'm surprised these matches don't cause compiler sadness, at least adding warnings logged. If they are, let's try to avoid that, we have enough logging noise in CI and local runs.

---

no warnings, but you are right that serializers should throw a nice exception when unexpected input. Fixed in https://github.com/akka/akka/pull/27882/commits/89b537f6e65aa82277fb7998fe0348a8e16ddce1